charmcraft/config.py

```diff
@@ -328,7 +335,22 @@ def unmarshal(cls, obj: Dict[str, Any], project: Project):
             # base configurations.  Doing it here rather than a Union
             # type will simplify user facing errors.
             bases = obj.get("bases")
-            if bases is not None and isinstance(bases, list):
+            if bases is None:
+                # Set default bases to Ubuntu 20.04 to match strict snap's
+                # effective behavior.
+                bases = [
+                    {
+                        "name": "ubuntu",
+                        "channel": "20.04",
+                        "architectures": [get_host_architecture()],
+                    }
+                ]
+                if not is_charmcraft_running_in_managed_mode():
+                    notify_deprecation("dn03")
+
+            # Expand short-form bases if only the bases is a valid list. If it
+            # is not a valid list, parse_obj() will properly handle the error.
+            if isinstance(bases, list):
```

Comment:
`bases` will always be a list, right?

Reply:
Should be, but this is prior to the proper error checking which would happen just after this.